python-sdk/rebuff/rebuff_pythonsdk.py

```diff
@@ -0,0 +1,228 @@
+import secrets
+from typing import Any, Optional, Tuple, Union
+from detect_pi_heuristics import detect_prompt_injection_using_heuristic_on_input
+from detect_pi_vectorbase import init_pinecone, detect_pi_using_vector_database
+from detect_pi_openai import render_prompt_for_pi_detection, call_openai_to_detect_pi
+from pydantic import BaseModel
+import langchain
+
+
+class Rebuff_Detection_Response(BaseModel):
+    heuristic_score: float
+    openai_score: float
+    vector_score: float
+    run_heuristic_check: bool
+    run_vector_check: bool
+    run_language_model_check: bool
+    max_heuristic_score: float
+    max_model_score: float
+    max_vector_score: float
+    injection_detected: bool
+
+
+class RebuffPython:
+    def __init__(
+        self,
+        openai_model: str,
+        openai_apikey: str,
+        pinecone_apikey: str,
+        pinecone_environment: str,
+        pinecone_index: str,
+    ) -> None:
+        self.openai_model = openai_model
+        self.openai_apikey = openai_apikey
+        self.pinecone_apikey = pinecone_apikey
+        self.pinecone_environment = pinecone_environment
+        self.pinecone_index = pinecone_index
+
+    def detect_injection(
+        self,
+        user_input: str,
+        max_heuristic_score: float = 0.75,
+        max_vector_score: float = 0.90,
+        max_model_score: float = 0.90,
+        check_heuristic: bool = True,
+        check_vector: bool = True,
+        check_llm: bool = True,
+    ) -> Rebuff_Detection_Response:
+        """
+        Detects if the given user input contains an injection attempt.
+
+        Args:
+            user_input (str): The user input to be checked for injection.
+            max_heuristic_score (float, optional): The maximum heuristic score allowed. Defaults to 0.75.
+            max_vector_score (float, optional): The maximum vector score allowed. Defaults to 0.90.
+            max_model_score (float, optional): The maximum model (LLM) score allowed. Defaults to 0.9.
+            check_heuristic (bool, optional): Whether to run the heuristic check. Defaults to True.
+            check_vector (bool, optional): Whether to run the vector check. Defaults to True.
+            check_llm (bool, optional): Whether to run the language model check. Defaults to True.
+
+        Returns:
+            Rebuff_Detection_Response
+        """
+
+        injection_detected = False
+
+        if check_heuristic:
+            rebuff_heuristic_score = detect_prompt_injection_using_heuristic_on_input(
+                user_input
+            )
+
+        else:
+            rebuff_heuristic_score = 0
+
+        if check_vector:
+            vector_store = init_pinecone(
+                self.pinecone_environment,
+                self.pinecone_apikey,
+                self.pinecone_index,
+                self.openai_apikey,
+            )
+
+            rebuff_vector_score = 0
+            similarity_threshold = 0.3
+            vector_store._text_key = "input"
```

Comment:
The initialization of Pinecone client requires a `_text_key` field.  I am not sure what is the `_text_key` exactly for though I have added the [reference](https://github.com/langchain-ai/langchain/blob/a6ebffb69504576a805f3b9f09732ad344751b89/langchain/vectorstores/pinecone.py#L57) as a comment in code as well.

Reply:
Ok, that link was helpful. `_text_key` is being use as the key for a metadata field that always gets added to a vector when it's inserted. If you don't specify it, it defaults to `"text"` as the metadata key. In the JS SDK, we're not specifying any value, so it's using the default value of `"text"`. Since we want Python and JS to match, I think you should remove this line.

Reply:
Thank you for elaborating on it, it makes more sense now. I think `_text_key` is also being used when [searching for similar documents](https://github.com/langchain-ai/langchain/blob/a6ebffb69504576a805f3b9f09732ad344751b89/langchain/vectorstores/pinecone.py#L126). Hence, if I remove the said line, I get the error: `Found document with no 'text' key. Skipping.`

@seanpmorgan can probably shed more light on it but I think the previous/earlier entries in Pinecone have been made against `self._text_key = 'input'`

Reply:
Ah, I see. In the particular Pinecone database you're testing against, the entries have a metadata field of `"input"` but not `"text"`, so the results all get skipped unless you set `self._text_key = 'input'`.

I understand why Langchain has this behavior (when querying a vector database, you almost always want the text content of the vector), but in our use case, we're not actually doing anything with the text content of the vector, just the score. I think the correct way forward at this point is to submit a PR to Langchain to have the option to not skip results when the metadata doesn't contain the text key (I'll submit that PR). I think you can still move forward with this PR in the meantime by instead of doing `self._text_key = 'input'`, do:
```py
vector_store = Pinecone.from_existing_index(index, openai_embeddings, text_key="input")
```
in `init_pinecone`. Whenever possible, it's better to set attributes via constructor or function arguments instead of directly on the instance (in Python, it's a convention that attributes that start with an underscore should not be accessed outside of that class).

I looked at the [JS code](https://github.com/langchain-ai/langchainjs/blob/8154c7829fc930a5b4060e830d06cbdd3e9484c4/libs/langchain-community/src/vectorstores/pinecone.ts#L230), and Langchain does not skip vectors when `this.textKey` is not found in the metadata, so we don't need to make any JS changes in regards to this issue.

Reply:
Langchain PR: https://github.com/langchain-ai/langchain/pull/15663

Reply:
Thank you for elaborating on it, and for submitting the Langchain PR. I have updated `init_pinecone` as you suggested.